### PR TITLE
Make current driver compatible with the OP_MSG protocol

### LINKED ISSUE #135: Support mongodb 6.0 ?
mongodb removed some opcodes in 6.0. that `mongodb-haskell` is using. See:
https://dochub.mongodb.org/core/legacy-opcode-removal

Are there plans on updating this driver so it supports 6.0?

### PR DESCRIPTION
**Pull request by kfiz**
Starting with mongodb v6 the OP_MSG protocol is the only accepted message protocol that is accepted by mongodb. All prior protocols are deprecated. This commit implements the protocol keeping the current client facing API intact.

Closes #135, #123

### CODE REVIEW

**Comment by kfiz** [comment]
@VictorDenisov Ping. Could you have a look at this? All tests have successfully passed on macOS x86_64 running ghc 9.2.4. This would need to be verified against older versions of mongoDB , older versions of ghc and on different architectures.

**Comment by VictorDenisov** [comment]
Yep. I'll take a look over the weekend.

**Comment by VictorDenisov** [comment]
It looks like it'll take longer than I expected. In progress ...

**Comment by VictorDenisov** [comment]
@kfiz if you make changes to the PR could you please add new commits. At least until we are done with this PR. We can squash them when we merge the PR.

**Comment by kfiz** [comment]
Ok. But I think I'm done with the changes for now. Hopefully this was not causing to much trouble.

File: Database/MongoDB/Query.hs [hunk 1]
```diff
@@ -1518,7 +1841,8 @@ type Command = Document
 
 runCommand :: (MonadIO m) => Command -> Action m Document
 -- ^ Run command against the database and return its result
-runCommand c = fromMaybe err <$> findOne (query c "$cmd") where
+runCommand c = do
+  fromMaybe err <$> findOne (query c "$cmd") where
```
Comment (VictorDenisov) [inline Database/MongoDB/Query.hs:1845]: nit: It looks like this code didn't really change. Let's keep only the necessary changes.

File: Database/MongoDB/Query.hs [hunk 2]
```diff
@@ -1527,7 +1851,12 @@ runCommand1 c = runCommand [c =: (1 :: Int)]
 
 eval :: (MonadIO m, Val v) => Javascript -> Action m v
 -- ^ Run code on server
-eval code = at "retval" <$> runCommand ["$eval" =: code]
+eval code = do
+    p <- asks mongoPipe
+    let sd = P.serverData p
+    if maxWireVersion sd <= 7
```
Comment (VictorDenisov) [inline Database/MongoDB/Query.hs:1857]: 7 or 17?

File: Database/MongoDB/Internal/Protocol.hs [hunk 3]
```diff
@@ -466,17 +736,38 @@ replyOpcode = 1
...
-    numDocs <- fromIntegral <$> getInt32
-    rDocuments <- replicateM numDocs getDocument
-    return (responseTo, Reply{..})
+    if opcode == 2013
+      then do
+            -- Notes:
+            -- Checksum bits that are set by the server don't seem to be supported by official drivers.
+            -- See: https://github.com/mongodb/mongo-python-driver/blob/master/pymongo/message.py#L1423
+            flagBits <-  rFlagsOpMsg <$> getInt32
+            _ ← getInt8
```
Comment (VictorDenisov) [inline Database/MongoDB/Internal/Protocol.hs:745]: It looks like you use a mix of UTF 8 chars and regular ascii chars. Could you please use only ascii characters?

On a personal note. What editor do you use to deal with them comfortably?

File: Database/MongoDB/Query.hs (hunk 2 above)
Comment (kfiz) [reply Database/MongoDB/Query.hs:1857]: 7 is correct. `eval` has been removed in mongoDB 4.2.

File: Database/MongoDB/Internal/Protocol.hs (hunk 3 above)
Comment (kfiz) [reply Database/MongoDB/Internal/Protocol.hs:745]: Yes, I usually use haskell's unicode support (enabled with  `{-# LANGUAGE UnicodeSyntax #-}`)  and I edit in emacs. Works great. But you're right I forgot to switch back to ascii. I'll push an update once I figured out how to do this efficiently.

File: Database/MongoDB/Query.hs (hunk 1 above)
Comment (kfiz) [reply Database/MongoDB/Query.hs:1845]: yep. will change that.

**Comment by VictorDenisov** [comment]
Looks good and good job keeping existing code intact. I'll merge your change tomorrow. It almost feels like we can only bump patch version. There was no change of external interfaces. Do you agree?

**Comment by kfiz** [comment]
Thanks. I agree, bumping the patch version is enough. No changes to external interfaces were made.

**Comment by averageflow** [comment]
@VictorDenisov @kfiz  and everyone involved, thanks for your awesome work. Maybe in the README it could be added a note specifying the version compatibility of MongoDB ? Do you also know what the process looks like for someone using Persistent lib + Mongo DB to use the latest versions?

**Comment by VictorDenisov** [comment]
I would expect little compatibility issues. No public interfaces have changed. Or are you talking about a different compatibility?

**Comment by kfiz** [comment]
> Do you also know what the process looks like for someone using Persistent lib + Mongo DB to use the latest versions?

I would assume that the Persistent maintainers would just have to make the latest version of this driver available as the default mongoDB backend. You would have to update your version of mongoDB of course if you haven't done that yet, if you want to try this out with mongoDB v6.0.

**Pull request merged**